[InstCombine] Canonicalize `switch(X^C)` expressions to `switch(X)`

antoniofrighetto opened the pull request:
`switch(X^C)` expressions can be folded to `switch(X)`. Minor opportunity to generalize simplifications in `visitSwitchInst` via an inverse function helper as well.

Proof: https://alive2.llvm.org/ce/z/TMRy_3.

Fixes: https://github.com/llvm/llvm-project/issues/174255.
Fixes: https://github.com/llvm/llvm-project/issues/143368.

antoniofrighetto commented:
> From llvm-opt-benchmark, a potential problem is that this can something turn small switch values into very large ones -- the common case is a xor by INT_MIN. I haven't checked how this affects codegen.

When not invoking the middle-end (except for SimplifyCFG, was not expecting llc to invoke it), the codegen looks slightly worse both in the optimized and unoptimized case: https://llvm.godbolt.org/z/jceqnev4n. Not sure if we wish to proceed further with the canonicalization :(

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -3878,26 +3878,29 @@ static Value *simplifySwitchOnSelectUsingRanges(SwitchInst &SI,
-      Case.setValue(cast<ConstantInt>(NewCase));
+  auto MaybeInvertible = [&](Value *Cond) -> std::optional<InvertFn> {
+    if (match(Cond, m_Add(m_Value(Op0), m_APInt(CondOpC))))
+      // Change 'switch (X+C) case A:' into 'switch (X) case C-A'.
```
```suggestion
      // Change 'switch (X+C) case A:' into 'switch (X) case A-C'.
```

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstructionCombining.cpp:
```diff
@@ -3878,26 +3878,29 @@ static Value *simplifySwitchOnSelectUsingRanges(SwitchInst &SI,
+
+  if (auto InvertFn = MaybeInvertible(Cond)) {
+    for (auto &Case : SI.cases()) {
+      const APInt &New = (*InvertFn)(Case.getCaseValue()->getValue(), *CondOpC);
```
```suggestion
      const APInt &New = (*InvertFn)(*CondOpC, Case.getCaseValue()->getValue());
```

dtcxzyw commented:
Do we need to check whether the condition is only used by the switch? Absorbing constants into switch cases may not be profitable if the condition is also used in other places...

antoniofrighetto commented:
> Do we need to check whether the condition is only used by the switch? Absorbing constants into switch cases may not be profitable if the condition is also used in other places...

Sounds reasonable to me.

antoniofrighetto commented:
@nikic Think we should proceed with this canonicalization? If so, should try `xor X, INT_MIN -> sub X, INT_MIN` first?

dtcxzyw reviewed (approved):
LG

nikic reviewed (commented):
Looks reasonable

antoniofrighetto committed:
[InstCombine] Canonicalize `switch(X^C)` expressions to `switch(X)`

`switch(X^C)` expressions can be folded to `switch(X)`. Minor
opportunity to generalize simplifications in `visitSwitchInst`
via an inverse function helper as well.

Proof: https://alive2.llvm.org/ce/z/TMRy_3.

Fixes: https://github.com/llvm/llvm-project/issues/174255.
Fixes: https://github.com/llvm/llvm-project/issues/143368.
